x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlExecutionInfo.java

```
@@ -168,8 +172,10 @@ public TimeValue planningTookTime() {
      */
     public void markEndQuery() {
         assert relativeStart != null : "Relative start time must be set when markEndQuery is called";
-        overallTimeSpan = relativeStart.stop();
-        overallTook = overallTimeSpan.toTimeValue();
+        if (isMainPlan()) {
```

Could you ever hit a scenario where markEndQuery is not called while isMainPlan() is true?, thus never finalizing overallTook?

Maybe you should throw an error if markEndQuery is called and isMainPlan is false?

---

I think that could be if the query fails? But then I guess we wouldn't expect the timings then? 

Calling `markEndQuery` in subplans wouldn't be an error - *their* parts of the query ended, so it's legit, just not the query we're interested in. I mean, it's of course possible to move the check to the call place, but I feel like it would be harder to read.